nRF5340: build TFM by default for the Non-secure version of the board

ioannisg opened the pull request:
This patch forces TFM as the default SPE solution for nRF5340.

It is the first Zephyr board that enables TF-M as the default SPE.

In particular, this PR:
- selects BUILD_WITH_TFM in the non-secure Zephyr builds for nRF5340, with TFM_BL2_FALSE=y (e.g. no MCUboot) built in.
- modifies the default 5340 SRAM partitioning to match what TFM is doing by default in flash layout and region definition.
- removes overlays from the integration samples (not needed)
- adds a release notes mention
- updates the documentation of nRF5340 DK to reflect that now TFM is the default SPE solution

ioannisg committed:
boards: arm: nrf5340: enable by default TF-M on non-secure builds

Enable building with TF-M by default on nRF5340 DK Application
core (cpuapp) when building for the non-secure version of the
board.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf5340: modify default SRAM partitioning

Modify the default partitioning of the Application core
SRAM, for Secure and Non-Secure domain, to accommodate
the default build configuration of TF-M. The RAM TF-M
uses should fit into the sram0_secure. The partitioning
should match what TF-M is allocating to secure and non-
secure domain.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf5340: correct ram information in board yml files

Re-adjust the available RAM advertized by the nRF5340 DK
and PDK .yml files (Application core, Non-Secure version
of the board).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
samples: tfm_integration: remove DTS overlays for nRF5340

Remove the DTS overlay files for nRF5340 DK and PDK,
since the SRAM partition changes are matching the
default SRAM partitioning in these boards.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

carlescufi commented on boards/arm/nrf5340dk_nrf5340/Kconfig.defconfig:
```diff
@@ -16,6 +16,12 @@ config BOARD
+# By default, if we build for a Non-Secure version of the board,
+# force building with TF-M as the Secure Execution Environment.
+
+config BUILD_WITH_TFM
```
Why do we want this by default exactly? Is there a reason why the current approach is not suitable?

ioannisg replied:
@carlescufi The upstream project, currently, does not have a solution for TEE on nRF5340; this patch is offering TFM as the default solution. For upstream users only.

This patch allows us to run full CI on the _NS board in the Zephyr project, using TFM as the SPE. Currently we have no coverage of _NS builds.

I do understand the implications of it in downstream use of the nRF5340 DK board definition, and I am open to suggestions. 

We can revert this `default y` if you think it is going to cause problems. Then building for _NS needs to manually enable TFM - and I don't see why, since there is no alternative..

I would keep, though the SRAM partition modifications, so we won't need to duplicate overlays everywhere.

tejlmand replied:
Just an observation.
Users who disable `BUILD_WITH_TFM` will still be restricted to only 192kB RAM in ns image, unless they create an overlay file themselves.

But is that clear to users ?

ioannisg replied:
> Users who disable BUILD_WITH_TFM will still be restricted to only 192kB RAM in ns image

@tejlmand thanks! So that's correct now: even without TF-M there will be less RAM available to the NS image.
Implications:
- **upstream**: nothing really. there is no upstream solution for TEE (other than tf-m) so if anyone wants to build for _NS, they will have to have their own SPE solution. If they are capable of doing that, i don't see why they cannot add an overlay with whatever SRAM they want.
- **downstream**: I believe NCS is using PM, so the default SRAM is anyways overwritten. So I don't see implications either. If there are NCS samples without PM, then I guess people will notice less SRAM in the NS side. Then they will need an overlay.

tejlmand replied:
Thanks for the explanation.

anangl commented on boards/arm/nrf5340dk_nrf5340/nrf5340_cpuapp_partition_conf.dts:
```diff
@@ -49,11 +49,11 @@
 
 &sram0_ns {
-	reg = <0x20010000 0x60000>;
+	reg = <0x20040000 0x30000>;
```
I think you need to also update this line accordingly:
https://github.com/zephyrproject-rtos/zephyr/blob/882d28b289d0c0d9053d6adffdb1b53a4fcc3849/boards/arm/nrf5340dk_nrf5340/nrf5340_cpuapp_common.dts#L196

ioannisg replied:
done

carlescufi reviewed (approved):
LGTM, thanks for the explanation

tejlmand reviewed (approved):
lgtm

jfischer-no commented on boards/arm/nrf5340dk_nrf5340/nrf5340dk_nrf5340_cpuappns.yaml:
```diff
@@ -6,7 +6,7 @@ toolchain:
   - xtools
   - zephyr
-ram: 384
+ram: 192
```
@ioannisg Should not [nrf5340dk_nrf5340_cpuapp.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/nrf5340dk_nrf5340/nrf5340dk_nrf5340_cpuapp.yaml#L9) also be adapted?

ioannisg replied:
Not really. That one should reflect _all_ memory that is available when we do not consider S, and NS builds. And this is already the case.

ioannisg committed:
doc: add comment about TF-M being the default SPE in nRF5340

Add a note in the 2.5 release notes, about TF-M being now
the default Secure Processing Element (SPE) for nRF5340 DK.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf5340: default to build TFM without BL2 for NS builds

When building with TF-M support (for non-secure Zephyr
applications) default to build TF-M without BL2 support.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: nrf5340: support TFM without BL2 for nRF5340 non-secure

Extend the nRF5340 board's CMakeLists.txt file to
support building TF-M without BL2 for nRF5340. The
result of the build is a single merged-hex containing
TF-M (SPE) and Zephyr (NSPE).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: nrf5340: update docs to reflect using TFM as the default SPE

Update the documentation of nRF5340 to stress that
TF-M is the default solution for building the Secure
image.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

carlescufi commented:
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/30501

ioannisg commented:
@PerMac FYI: 
I verified arch/arm and kernel/ tests execute properly on nRF5340 CPUAPP Non-Secure (no failures related to the Non-secure domain)

You might want to add the nrf5340dk_nrf5340_cpuappns board into the list of nRF platforms being tested.
